doc/src/tutorial/codegen.rst

```diff
@@ -0,0 +1,516 @@
+===============
+Code Generation
+===============
+Several submodules in Sympy are provided to generate directly compilable 
+code from Sympy expressions. Furthermore, there are submodules provided 
+that help use other numeric systems as backend for construction and 
+manipulation of mathematical expressions to speed up the process.
+
+We will start with a brief introduction, and inspect the building 
+blocks of the code generation and then the code generation itself.
+
+Introduction
+------------
+
+There are three levels::
+
+    autowrap
+       |
+    codegen
+       |
+    code printers
+
+Autowrap uses codegen, and codegen uses the code printers. Autowrap does 
+everything: it lets you go from SymPy expression to numerical 
+function in the same Python process in one step. codegen is actual 
+code generation, i.e., to compile and use later, or to include in some larger 
+project.
+
+The code printers translate the SymPy objects into actual code, 
+like abs(x) -> fabs(x) (for C).
+
+The code printers don't print optimal code in many cases. 
+An example of this is powers in C. x**2 prints as pow(x, 2) instead of x*x. 
+Other optimizations (like mathematical simplifications) should happen 
+before the code printers.
+
+cse is not applied anywhere in this chain. It ideally happens at the 
+codegen level, or somewhere above it.
+
+We will iterate through the levels, bottom up.
+
+Code printers (sympy.printing)
+------------------------------
+This is where the meat of code generation is, the translation of SymPy
+expressions to specific code.Supported languages are C (``ccode``), Fortran 95 (``fcode``), 
+Javascript (``jscode``), Mathematica (``mathematica_code``), Octave/Matlab (``octave_code``), 
```

Comment:
`jscode`, `octave_code`, `theano_code` aren't rendering properly, because they haven't been included in `printing.rst` doc in modules. What should be done about them?